R/vec-coerce.R

```diff
@@ -0,0 +1,172 @@
+#'
+#' * Have symmetric coercion behaviour. The resulting type should be
+#'   identical on permutation of `from` and `to`.
+#'
+#' * Not necessarily be fully commutative: `vec_coerce()` is expected
+#'   to return a vector as long as the first argument `from`.
+#'
+#' * Differentiate between empty and non-empty vectors. Most of the
+#'   time all the type information should be contained in attributes
+#'   and the return type can be determined on empty vectors.
+#'
+#'   In some cases however the contents of the vector are necessary to
+#'   determine the appropriate coercion (e.g. the contents of a
+#'   character vector are needed to determine congruence with the
+#'   levels of a factor).
+#'
+#' * Should warn about dangerous coercions with the class
+#'   `"rlang_vec_coerce_wng"` so that coercion warnings can be
+#'   selectively muffled (see [muffle_vec_coerce()]).
+#'
+#' @param from The vector to coerce. The return value should be a
+#'   vector as long as `from`.
+#' @param to The vector containing the type information to determine
+#'   the coercion of `from`.
+#'
+#' @include dispatch.R compat-purrr.R
+#' @export
+vec_coerce <- function(from, to) {
+  dispatch2("vec_coerce", from, to)
+}
+
+## Export all methods
+#' @rawNamespace export(.__rlang_binary_strict_methods__.)
+NULL
+
+
+### Numeric coercions
+
+def_method2("logical", "logical",
+  vec_coerce = function(from, to, ...) {
+    from
+  }
+)
+def_method2("logical", "integer",
+  vec_coerce = function(from, to, ...) {
+    vec_coerce_bare(from, "integer")
+  }
+)
+def_method2("logical", "numeric",
+  vec_coerce = function(from, to, ...) {
+    vec_coerce_bare(from, "numeric")
+  }
+)
+
+def_method2("integer", "integer",
+  vec_coerce = function(from, to, ...) {
+    from
+  }
+)
+def_method2("integer", "numeric",
+  vec_coerce = function(from, to, ...) {
+    vec_coerce_bare(from, "numeric")
+  }
+)
+
+def_method2("numeric", "numeric",
+  vec_coerce = function(from, to, ...) {
+    from
+  }
+)
+
+
+### Character coercions
+
+def_method2("character", "character",
+  vec_coerce = function(from, to, ...) {
+    from
+  }
+)
+
+def_method2("character", "factor",
+  vec_coerce = function(from, to, ...) {
+    chr <- .dispatched$character
+    fct <- .dispatched$factor
+
+    if (!length(chr) || !all(chr %in% levels(fct))) {
+      msg <- "Coercion of `factor` to `character` loses levels information"
+      warn(msg, "rlang_vec_coerce_wng")
+      return(as.character(from))
+    }
+
+    factor(from, levels(fct))
+  }
+)
+
+def_method2("factor", "factor",
+  vec_coerce = function(from, to, ...) {
+    lvls_from <- levels(from)
+    lvls_to <- levels(to)
+
+    if (!(lvls_from %in% lvls_to) && !(lvls_to %in% lvls_from)) {
+      msg <- "Coercing `factor` to `character` because of incompatible levels"
+      warn(msg, "rlang_vec_coerce_wng")
+      return(as.character(from))
+    }
+
+    n_from <- length(lvls_from)
+    n_to <- length(lvls_to)
+    if (n_from != n_to) {
```

Comment:
That sounds correct; I just find the current logic in your code hard to follow.

Reply:
sorry, how would you do it?

Reply:
I think I'd work in the opposite direction:

```R
if (setequal(lvls_from, lvls_to)) {
  factor(from, lvls_to)
} else if (all(lvls_from %in% lvls_to)) {
  warn(...)
  factor(from, lvls_to)
} else {
  warn(...)
  as.character(from)
}
```

Reply:
Thanks it's much better. I'd just always use the longer levels rather than `lvls_to` in the congruent case so that coercion behaviour is symmetric (and it's the choice that preserves the most information).

Reply:
Well actually that's not right: `all(lvls_from %in% lvls_to)` is FALSE when LHS is longer than rhs. How about:

```r
    union <- union(lvls_from, lvls_to)
    lengths <- c(length(lvls_from), length(lvls_to))
    congruent <- length(union) == lengths

    if (all(congruent)) {
      from
    } else if (any(congruent)) {
      msg <- "Factor levels are congruent but not the same length"
      warn(msg, "rlang_vec_coerce_wng")
      lvls <- if (lengths[[1]] > lengths[[2]]) lvls_from else lvls_to
      factor(from, lvls)
    } else {
      msg <- "Coercing `factor` to `character` because of incompatible levels"
      warn(msg, "rlang_vec_coerce_wng")
      as.character(from)
    }
```